pkg/scheduler/preemption/fairsharing/least_common_ancestor.go

```diff
@@ -0,0 +1,59 @@
+// Copyright The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fairsharing
+
+import "sigs.k8s.io/kueue/pkg/cache"
+
+// almostLCA is defined on two ClusterQueues, as the two nodes before
+// the lowest shared node - the LeastCommonAncestor (LCA). While LCA
+// is always a Cohort, almostLCA may be a ClusterQueue or a Cohort.
+type almostLCA interface {
+	DominantResourceShare() int
+}
+
+// getAlmostLCAs returns almostLCAs of (preemptor, target).
+func getAlmostLCAs(t *TargetClusterQueue) (almostLCA, almostLCA) {
+	lca := getLCA(t)
+	return getAlmostLCA(t.ordering.preemptorCq, lca), getAlmostLCA(t.targetCq, lca)
+}
+
+// getLCA traverses from a ClusterQueue towards the root Cohort,
+// returning the first Cohort which contains the preemptor
+// ClusterQueue in its subtree.
+func getLCA(t *TargetClusterQueue) *cache.CohortSnapshot {
+	cohort := t.targetCq.Parent()
+	for {
+		if t.ordering.onPathToPreemptorCQ(cohort) {
+			return cohort
+		}
+		cohort = cohort.Parent()
+	}
```

Comment:
Seeing the pattern repeated in a few places already  I would really like to have an iterator, like say `(c *ClusterQueue) pathToRoot() seq.Iter[]`, over the path to the root. Then, we could just for-range in here IIUC.

Still, it might be a follow up cleanup issue / pr.

Reply:
Created #4644, as I think it would be best to solve all the cases at once

Reply:
Oh I also opened: https://github.com/kubernetes-sigs/kueue/issues/4645 :) such a race condition :)

Reply:
closed mine as you were first